File: src/xal/sim/slg/Lattice.py
# $Id $
############################################################################
import sys
from math import fabs
############################################################################
LatticeError="LATTICE_ERROR:"
true =(1>0)  #get true
false=(1<0)  #get false
eps=1.e-5    #precision limit for position calculations
debug=false
#.........................................................
class Lattice:
	"""The lattice is a list of elements. Each element is associated with one and only one acclerator node."""

	def __init__(self,name,base=0):
		"""Create an empty lattice with a 'name' and a 'base' position."""
		self.name=name
		self.base=base    #the base position of the latice
		self.elements=[]  #the element list
		#an empty lattice
		self.elements.append(Marker(0,0,"BEGIN "+name))
		self.elements.append(Marker(0))
		self.elements.append(Marker(0,0,"END "+name))
		#all elements get same offset
		self.updateBases()

	def __len__(self):
		"""Return the lattice length in number of elements."""
		return len(self.elements)

	def __getitem__(self,index):
		"""Get element with 'index' from lattice."""
		return self.elements[index]

	def getLength(self):
		"""Return the lattice length in distance units."""
		return self.elements[-1].getEndPosition()

	def getName(self):
		"""Return the lattice name."""
		return self.name

	def append(self,element):
		"""Append an element at the end of the lattice."""
		lattice_end=self.elements.pop()
		last=self.elements[-1]
		start_pos=element.getStartPosition()
		#space to fill up with drift space
		drift_len=start_pos-last.getPosition()
		if drift_len < -eps:
			#ooops negative length: severe error
			raise LatticeError,"length is negative when appending: "+element.getName()
		elif fabs(drift_len) < eps:
			#too short drift: ignore
			pass
		else:
			#add an upstream drift space
			drift_pos=last.getPosition()+drift_len*0.5
			self.elements.append(Drift(drift_pos,drift_len))
			self.elements.append(Marker(start_pos))
		#add the element
		self.elements.append(element)
		end_pos=element.getEndPosition()
		self.elements.append(Marker(end_pos))
		#place 'lattice END' marker
		self.elements.append(Marker(end_pos,0,lattice_end.getName()))
		#all elements get same offset
		self.updateBases()
	
	def appendTuple(self,tuple):
		"""Append a tuple to the end of the lattice"""
		for el in tuple:
			self.append(el)

	def insert(self,element):
		"""Insert a zero length element into the lattice. Lumped elements in the lattice will be sliced accordingly."""
		if element.getLength() != 0.0:
			raise LatticeError,"length must be zero when inserting: "+element.getName()
		#assemble a list of indexes of marker elements
		markers=[]
		for i in range(len(self)):
			if self.elements[i].type == "marker":
				markers.append(i)
		#search the markers that embrace the element 
		before=markers[0]
		after=markers[-1]
		for m in markers[1:-1]:
			if self.elements[m].getPosition() <= element.getPosition():
				before=m
				continue
			after=m
			break
		#slice the element between the two markers
		between=after-1
		self.elements[between:after]=self.elements[between].split(element)
		#all elements get same offset
		self.updateBases()

	def clearMarkers(self):   
		"""Remove all marker elements from the lattice."""
		lattice=[]
		begin=self.elements[0]
		end  =self.elements[-1]
		lattice.append(begin)
		for el in self.elements:
			if el.getType() == 'marker': continue
			lattice.append(el)
		lattice.append(end)
		self.elements=lattice

	def updateBases(self):
		"""NOT FOR PUBLIC USE"""
		#all elements get same offset
		for el in self.elements:
			el.setBase(self.base)

	def joinDrifts(self):
		"""Join neighboring drift spaces into a single one."""
		lattice=[]
		#pass 1
		end=self.elements[-1]
		ix=0
		while ix < len(self)-1:
			el=self.elements[ix]
			ix=ix+1
			if el.getType() == 'drift': 
				elnext=self.elements[ix]
				if elnext.getType() == 'drift':
					ix=ix+1
					len1=el.getLength()
					len2=elnext.getLength()
					spos1=el.getStartPosition()
					spos2=elnext.getStartPosition()
					jlen=len1+len2
					jpos=spos1+jlen*0.5
					jspos=jpos-jlen*0.5
					jdrift=Drift(jpos,jlen)
					lattice.append(jdrift)
					if debug: print "join drifts:",(spos1,len1),"+",(spos2,len2),"=",(jspos,jlen)
				else:
					lattice.append(el)
			else: 
				lattice.append(el)
		lattice.append(end)
		self.elements=lattice

		#pass 2
		#check if there are still touching drifts in the lattice
		#if so call this function recursively until all done
		ix=0
		while ix < len(self)-1:
			el=self.elements[ix]
			ix=ix+1
			if el.getType() == 'drift':
				elnext=self.elements[ix]
				if elnext.getType() == 'drift':
					#repeat the joining process
					self.joinDrifts()
					return
				else:
					pass
			else:
				pass
		return

	def cout(self):
		"""Make a printed output of the lattice."""
		for count in range(len(self)):
			el=self.elements[count]
			abs_start_pos=el.getAbsPosition()-el.getLength()*0.5
			print "s=",abs_start_pos,"m\t",el.getName(),"\t",el.getType(),el.getPosition(),el.getLength()
		print "Totals: length of",self.getName(),"=",self.getLength(),"m with",len(self),"elements."

	def isConsistent(self):
		"""Consistency check"""
		for ix in range(len(self)-1):
			el=self[ix]
			next=self[ix+1]
			el_pos=el.getPosition()
			el_len=el.getLength()
			next_pos=next.getPosition()
			next_len=next.getLength()
			if fabs(el_pos+(el_len+next_len)*0.5-next_pos) > eps:
				raise LatticeError,"inconsistent distances between "+el.getName()+" and "+next.getName()
#.........................................................
class Element:
	"""The super class of all lattice elements."""

	def __init__(self,name,position=0,len=0):
		"""NOT FOR PUBLIC USE"""
		self.name=name
		self.position=position
		self.base=0
		self.len=len

	def getStartPosition(self):
		"""Return the upstream start position of this element."""
		return self.position-self.len*0.5

	def getEndPosition(self):
		"""Return the downstream end position of this element."""
		return self.position+self.len*0.5

	def getPosition(self):
		"""Return the center position of this element."""
		return self.position

	def getAbsPosition(self):
		"""Return the absolute center position of this element."""
		return self.position+self.base

	def getLength(self):
		"""Return the length of the element in distance units."""
		return self.len

	def getName(self):
		"""Return the name of the element."""
		return self.name

	def getBase(self):
		"""Return the base for relative positions."""
		return self.base

	def getSlicePositions(self,cut_pos):
		"""NOT FOR PUBLIC USE"""
		#calculate length and position of sliced parts
		up_len=cut_pos-self.getStartPosition()
		if fabs(up_len) < eps: up_len = 0
		dn_len=self.getLength()-up_len
		if fabs(dn_len) < eps: dn_len = 0
		up_pos=self.getStartPosition()+up_len*0.5
		dn_pos=self.getEndPosition()-dn_len*0.5
		return (up_pos,up_len,dn_pos,dn_len)

	def setBase(self,base):
		"""NOT FOR PUBLIC USE"""
		#Set the base for relative positions.
		self.base=base

	def setPosition(self,position):
		"""Set the element position."""
		self.position=position

	def setLength(self,length):
		"""Set the element length."""
		self.len=length

	def split(self,insert,CLASS):
		"""NOT FOR PUBLIC USE"""
		#The slice (and replace) operation. The thick element 'self' of type 'CLASS'
		#is cut into an upstream and a downstream part and then element 'insert'
		#is inserted (with limiting markers) into the lattice.
		cut_pos=insert.getPosition()
		positions=self.getSlicePositions(cut_pos)
		upstream=CLASS(positions[0],positions[1],self.getName())
		dnstream=CLASS(positions[2],positions[3],self.getName()+"+")
		marker=Marker(cut_pos)
		if fabs(upstream.getLength()) < eps:
			return [insert,marker,self]
		elif fabs(dnstream.getLength()) < eps:
			return [self,marker,insert]
		else:
			return [upstream,marker,insert,marker,dnstream]
#.........................................................
class Drift(Element):
	"""The drift space element."""
	type="drift"

	def __init__(self,position,len,name="---"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,Drift)
#.........................................................
class Dipole(Element):
	"""The dipole element."""
	type="dipole"

	def __init__(self,position,len,name="DIP"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,Dipole)
#.........................................................
class Quadrupole(Element):
	"""The quadrupole element."""
	type="quadrupole"

	def __init__(self,position,len,name="NQP"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,Quadrupole)
#.........................................................
class Sextupole(Element):
	"""The sextupole element."""
	type="sextupole"

	def __init__(self,position,len,name="NSX"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,Sextupole)
#.........................................................
class SkewQuad(Element):
	"""The skew quadrupole element."""
	type="skew_quadrupole"

	def __init__(self,position,len,name="SQP"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,SkewQuad)
#.........................................................
class SkewSext(Element):
	"""The skew sextupole element."""
	type="skew_sextupole"

	def __init__(self,position,len,name="SSX"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,SkewSext)
#.........................................................
class Octupole(Element):
	"""The octupole element."""
	type="octupole"

	def __init__(self,position,len,name="OCT"):
		"""Create the element: center position and len(ght) are needed"""
		Element.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return Element.split(self,insert,Octupole)
#.........................................................
class ThinElement(Element):
	"""The superclass of all thin elements"""

	def __init__(self,name,position=0,len=0):
		"""NOT FOR PUBLIC USE"""
		Element.__init__(self,name,position,len)

	def getStartPosition(self):
		"""Return the upstream start position of this element."""
		return Element.getPosition(self)

	def getEndPosition(self):
		"""Return the downstream end position of this element."""
		return Element.getPosition(self)

	def getLength(self):
		"""Return the length of the element in distance units."""
		return 0

	def getEffLength(self):
		"""Return the effective length of the element in distance units."""
		return Element.getLength(self)

	def getUpstreamDrift(self):
		"""Return the upstream drift space."""
		len=self.getEffLength()*0.5
		position=Element.getPosition(self)
		if fabs(len) < eps:
			return Marker(position)
		else:
			position=position-len*0.5
			return Drift(position,len)

	def getDownstreamDrift(self):
		"""Return the downstream drift space."""
		len=self.getEffLength()*0.5
		position=Element.getPosition(self)
		if fabs(len) < eps:
			return Marker(position)
		else:
			position=position+len*0.5
			return Drift(position,len)

	def asTuple(self):
		"""Return the thin element as a tuple (drift,element,drift)"""
		return (self.getUpstreamDrift(),self,self.getDownstreamDrift())

	def split(self,insert):
		"""NOT FOR PUBLIC USE"""
		#Split this element to place element 'insert' in between.
		return [self,insert]
#.........................................................
class Marker(ThinElement):
	"""The marker element."""
	type="marker"

	def __init__(self,position,len=0,name="***"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
class RFGap(ThinElement):
	"""The radio frequency gap element."""
	type="rfgap"

	def __init__(self,position,len=0,name="RFG"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
class BCMonitor(ThinElement):
	"""The beam current monitor element."""
	type="beamcurrentmonitor"

	def __init__(self,position,len=0,name="BCM"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position,len)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
class HSteerer(ThinElement):
	"""The horizontal steerer element."""
	type="hsteerer"

	def __init__(self,position,len=0,name="DCH"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
class VSteerer(ThinElement):
	"""The vertical steerer element."""
	type="vsteerer"

	def __init__(self,position,len=0,name="DCV"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
class BPMonitor(ThinElement):
	"""The beam position monitor element."""
	type="beampositionmonitor"

	def __init__(self,position,len=0,name="BPM"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
class WScanner(ThinElement):
	"""The wire scanner monitor element."""
	type="wirescanner"

	def __init__(self,position,len=0,name="WSM"):
		"""Create the element: center position and len(ght) are needed"""
		ThinElement.__init__(self,name,position)

	def getType(self):
		"""Return the element type."""
		return self.type;
#.........................................................
if __name__ == '__main__':
	test="Thick Element Test"
	print test
	lattice=Lattice(test,40.)
	#..............................
	try:
		lattice.append(Dipole(0.5,0.3,"DIP/0"))
		lattice.append(Quadrupole(0.85,0.15,"QH/0"))
		#lattice.append(Quadrupole(1.1,0.15,"QV/3"))
		lattice.append(Quadrupole(1.1,0.15))
		lattice.append(Dipole(1.37,0.3,"DIP/1"))
		#lattice.append(BPMonitor(1.57))
		lattice.append(Drift(1.545,0.05))
		#..............................
		#lattice.insert(BPMonitor(0.0,0,">>>"))
		#lattice.insert(BPMonitor(0.5,0,">>>"))
		#lattice.insert(BPMonitor(1.1,0,">>>"))
		#lattice.insert(BPMonitor(1.1375,0,">>>"))
		#lattice.insert(BPMonitor(1.175,0,">>>"))
		#lattice.insert(BPMonitor(1.57,0,">>>"))
		#..............................
		#lattice.insert(VSteerer(0.0))
		#lattice.insert(HSteerer(0.0))
		#lattice.insert(VSteerer(lattice.getLength()))
		#lattice.insert(HSteerer(lattice.getLength()))
		#..............................
		#lattice.clearMarkers()
		lattice.cout()
		#..............................
		test1="Thin Element Test"
		print test1
		lattice1=Lattice(test1,66)
		rfgap=RFGap(10,20)
		for part in rfgap.asTuple():
			start_pos=part.getStartPosition()
			name=part.getName()
			type=part.getType()
			position=part.getPosition()
			length=part.getLength()
			print start_pos,name,type,position,length	
			lattice1.append(part)
		lattice1.clearMarkers()
		lattice1.cout()
		#..............................
		lattice2=Lattice(test1,66)
		rfgap=RFGap(45.3,0)
		for part in rfgap.asTuple():
			start_pos=part.getStartPosition()
			name=part.getName()
			type=part.getType()
			position=part.getPosition()
			length=part.getLength()
			print start_pos,name,type,position,length	
			lattice2.append(part)
		lattice2.clearMarkers()
		lattice2.cout()
		#..............................
	except LatticeError,message:
		print LatticeError,message
		sys.exit(-1)

File: test/core/src/xal/sim/slg/Lattice.java
/*
 * Lattice.java
 *
 * Created on March 16, 2003, 8:58 PM
 */

package xal.sim.slg;

import java.io.PrintStream;
import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.*;
import org.w3c.dom.*;
/**
 * The lattice is a linear sequence of elements. Each element is associated
 * with one and only one XAL AcceleratorNode object.
 *
 * @author  wdklotz
 */
public class Lattice implements Cloneable {
	private String name; 																//the lattice name
	private double base;															//the base position
	private List<Element> elements; 													//the element list
	private PrintStream cout; 														//console
	public static final double EPS; 											//precision limit for position calculations
	private boolean debug; 														// print flag
	private boolean verbose; 													// print flag
	public static NumberFormat fmt; 										//number formater
	private Node2ElementMapper node2ElementMap; 				//dictionary [(key,value)=(node,element)]

	static {
		fmt= NumberFormat.getNumberInstance();
		((DecimalFormat) fmt).applyPattern("0.000000");
		EPS= 1.e-5d;
	}

	/** Creates an empty lattice with a 'name' and a 'base' position.
	 * @param name the lattice name.
	 * @param base the lattice base position.
	 *  */
	public Lattice(String name, double base) {
		this.name= name;
		this.base= base;
		elements= new ArrayList<Element>();
		cout= System.out;
		debug= false;
		verbose= false;
		//an empty lattice
		elements.add(new PermMarker(0.d, 0.d, "BEGIN_" + name));
		elements.add(new Marker(0.d));
		elements.add(new PermMarker(0.d, 0.d, "END_" + name));
		//all elements get same offset
		updateBases();
	}

	/** Creates an empty lattice with a 'name' and a 'base' position */
	public Lattice(String name, Double base) {
		this(name, base.doubleValue());
	}

	/** Creates an empty lattice with a 'name' and a 'base=0' position */
	public Lattice(String name) {
		this(name, 0.d);
	}

	/**
	 * Return the lattice base position.
	 */
	public double getBase() {
		return base;
	}

	/**
	 * Set the lattice base position.
	 */
	public void setBase(double base) {
		this.base= base;
		updateBases();
	}

/**Clone a lattice. The cloned lattice is a deep copied clone of the original.*/
	public Object clone() {
		try {
			Lattice cloned= (Lattice) super.clone();
			//make a deep copy of the list of lattice elements
			cloned.elements= new ArrayList<Element>();
			LatticeIterator liter= latticeIterator();
			while (liter.hasNext()) {
				Element elm= (Element) liter.next().clone();
				cloned.elements.add(elm);
			}
			return cloned;
		} catch (CloneNotSupportedException e) {
			//cannot happen -- we support clone
			throw new InternalError(e.toString());
		}
	}
	
	/**
	 * Concatenate two lattices. The resulting lattice will be a lattice with right added to the end of
	 * left. 
	 * @return the combined lattice from left to right.
	 * */
	static Lattice concatenate(Lattice left, Lattice right) throws LatticeError {
		Lattice part1 = (Lattice) left.clone();
		Lattice part2 = (Lattice) right.clone();
		part2.setBase(part1.getLength());    //shift position base of second part
		Lattice part3 = new Lattice(part1.getName()+"+"+part2.getName(),0d);
		part3.clearMarkers(); //remove dummy makers
		int last = part3.len()-1;
		part3.elements.addAll(last,part1.elements);    //add all from 1st part
		last = part3.len()-1;
		part3.elements.addAll(last,part2.elements);  //add all from 2nd part
		last = part3.len()-1;
		//update position of END marker
		(part3.elements.get(last)).setPosition(part1.getLength()+part2.getLength());
		return part3;
	}

	/**
	 * Return the lattice length in number of elements.
	 */
	public int len() {
		return elements.size();
	}

	/**
	 * Get element at 'index' from lattice.
	 */
	public Element getItem(int index) {
		return elements.get(index);
	}

	/**
	 * Return the lattice length in distance units.
	 */
	public double getLength() {
		return getItem(len() - 1).getEndPosition();
	}

	/**
	 * Set the lattice name property.
	 */
	public void setName(java.lang.String name) {
		this.name= name;
	}

	/**
	 * Return the name property.
	 */
	public String getName() {
		return name;
	}

	/**
	 * Setter for the forward dictionary property: [(key,value)=(node,element)]
	 */
	public void setNode2ElementMapper(Node2ElementMapper node2ElementMap) {
		this.node2ElementMap= node2ElementMap;
	}

	/**
	 * Getter for the forward dictionary property: [(key,value)=(node,element)]
	 */
	public Node2ElementMapper getNode2ElementMapper() {
		if (node2ElementMap == null) {
			//make & deposit the accumulated map in the lattice object
			setNode2ElementMapper(makeNode2ElementMapper());
		}
		return node2ElementMap;
	}

	/**
	 *  Make a new forward dictionary: [(key,value)=(node,element)].
	 * @return The forward dictionary that maps accelerator nodes to lattice elements.
	 */
	public Node2ElementMapper makeNode2ElementMapper() {
		//generate the forward dictionary,i.e. [(key,value)=(node,element)]
		Node2ElementMapper mapper= new Node2ElementMapper();
		LatticeIterator liter= latticeIterator();
		while (liter.hasNext()) {
			// visit every lattice element
			liter.next().accept((Visitor) mapper);
		}
		return mapper;
	}

	/**
	 * Append an element at the end of the lattice.
	 */
	void append(Element element) throws LatticeError {
		if (debug) {
			cout.println(
				"append: " + element.getName() + ", pos= " + element.getPosition() + ", len= " + element.getLength());
		}
		Element lattice_end= elements.remove(len() - 1);
		Element last= getItem(len() - 1);
		double start_pos= element.getStartPosition();
		//is there space to fill up with drift space ?
		double drift_len= start_pos - last.getPosition();
		if (drift_len < -EPS) {
			//ooops! negative length: severe error ...
			String message=
				"negative length when appending: " + element.getName() + ": calculated length= " + drift_len;
			message += "\nlast: " + last.getName() + ": pos= " + last.getPosition() + ", len= " + last.getLength();
			throw new LatticeError(message);
		} else if (Math.abs(drift_len) < EPS) {
			/*too short drift: ingnore*/;
		} else {
			//add an upstream drift space
			double drift_pos= last.getPosition() + drift_len * 0.5;
			elements.add(new Drift(drift_pos, drift_len));
			elements.add(new Marker(start_pos));
		}
		//add the element
		elements.add(element);
		double end_pos= element.getEndPosition();
		elements.add(new Marker(end_pos));
		//place lattice 'END' marker
		elements.add(new PermMarker(end_pos, 0.0, lattice_end.getName()));
		//all elements get same offset
		updateBases();
	}

	/**
	 * Append a tuple to the lattice end.
	 */
	void appendTuple( final List<Element> tuple ) throws LatticeError {
		for ( final Element element : tuple ) {
			append( element );
		}
	}

	/**
	 * Insert a thin element (length=0) into the lattice. Thick elements
	 * at the same position will be sliced accordingly.
	 */
	void insert(Element element) throws LatticeError {
		if (element.getLength() != 0.0) {
			throw new LatticeError("length must be zero when inserting: " + element.getName());
		}
		//assemble a list of indices of marker elements
		//exclude BEGIN and END markers
		int[] markers= new int[3 * len()];
		int jx= 0;
		for (int ix= 1; ix < len() - 1; ix++) {
			if (getItem(ix).getType().equals("marker")) {
				markers[jx++]= ix;
			}
		}
		//        if(false) {
		//            for(int ix=0; ix<jx; ix++) {
		//                cout.println("makers["+ix+"]="+markers[ix]+": p="+getItem(markers[ix]).getPosition());
		//            }
		//        }
		//search for markers that embrace the element
		int before= markers[0];
		int after= markers[jx - 1];
		for (int m= 0; m < jx; m++) {
			if (getItem(markers[m]).getPosition() <= element.getPosition()) {
				before= markers[m];
				continue;
			}
			after= markers[m];
			break;
		}
		//slice the element between the two markers
		int between= after - 1;
		//        if(false) {cout.println("(before,between,after)=("+before+", "+between+", "+after+")");}
		Element to_split= elements.remove(between);
		final List<Element> to_insert= to_split.split(element);
		elements.addAll(between, to_insert);
		if (debug & verbose) {
			cout.println(
				"insert: replacing "
					+ to_split.getName()
					+ ", p= "
					+ to_split.getPosition()
					+ ", l= "
					+ to_split.getLength()
					+ " with");
			ListIterator<Element> lit= to_insert.listIterator();
			while (lit.hasNext()) {
				Element el= lit.next();
				cout.println("\t" + el.getName() + ", p= " + el.getPosition() + ", l= " + el.getLength());
			}
		}
		//all elements get same offset
		updateBases();
	}

	/**
	 * Remove all non permanent marker elements from the lattice.
	 */
	public void clearMarkers() {
		for (LatticeIterator lit= latticeIterator(); lit.hasNext();) {
			if (lit.next().getType().equals("marker")) {
				lit.remove();
			}
		}
	}

	/**
	 * All elements get the same offset.
	 */
	private void updateBases() {
		for (LatticeIterator lit= latticeIterator(); lit.hasNext();) {
			lit.next().setBase(base);
		}
	}

	/**
	 * Join neighbouring drift spaces into a single one.
	 */
	public void joinDrifts() {
		ArrayList<Element> lattice= new ArrayList<Element>();
		//pass1
		Element end= getItem(len() - 1);
		int ix= 0;
		while (ix < len() - 1) {
			Element el= getItem(ix);
			ix += 1;
			if (el.getType().equals("drift")) {
				Element elnext= getItem(ix);
				if (elnext.getType().equals("drift")) {
					ix += 1;
					double len1= el.getLength();
					double len2= elnext.getLength();
					double spos1= el.getStartPosition();
					double spos2= elnext.getStartPosition();
					double jlen= len1 + len2;
					double jpos= spos1 + jlen * 0.5;
					double jspos= jpos - jlen * 0.5;
					lattice.add(new Drift(jpos, jlen));
					if (debug) {
						cout.println(
							"join drifts: ("
								+ spos1
								+ ","
								+ len1
								+ ") + ("
								+ spos2
								+ ","
								+ len2
								+ ") = ("
								+ jspos
								+ ","
								+ jlen
								+ ")");
					}
				} else {
					lattice.add(el);
				}
			} else {
				lattice.add(el);
			}
		}
		lattice.add(end);
		elements= lattice;

		//pass 2
		//check if there are still touching drifts in the lattice,
		//if so call this function recursively until all done
		ix= 0;
		while (ix < len() - 1) {
			Element el= getItem(ix);
			ix += 1;
			if (el.getType().equals("drift")) {
				Element elnext= getItem(ix);
				if (elnext.getType().equals("drift")) {
					//repeat the joining process
					joinDrifts();
					return;
				}
			}
		}
	}

	//	public static Lattice concatenate(Lattice first, Lattice second) {
	//		Lattice result = new Lattice(first.getName()+"+"+second.getName());
	//		result.appendTuple(first.elements);
	//		result.appendTuple(second.elements);
	//		return result;
	//	}

	/**
	 * Make a printed output of the lattice.
	 */
	void toConsole() {
		for (LatticeIterator lit= latticeIterator(); lit.hasNext();) {
			cout.println(lit.next().toCoutString());
		}
		cout.println("Totals: length of " + getName() + " = " + getLength() + " m, with " + len() + " elements.\n");
	}

	/**
	 * A consistency check.
	 */
	void forConsistency() throws LatticeError {
		for (int ix= 0; ix < len() - 1; ix++) {
			Element el= getItem(ix);
			Element next= getItem(ix + 1);
			double el_pos= el.getPosition();
			double el_len= el.getLength();
			double next_pos= next.getPosition();
			double next_len= next.getLength();
			if (Math.abs(el_pos + (el_len + next_len) * 0.5 - next_pos) > EPS) {
				//ooops! this should never happen, continue debugging.
				throw new LatticeError("consistency check failed at about " + el.getName() + " and " + next.getName());
			}
		}
	}

	/*
	 * Debugging output.
	 */
	void setDebug(boolean debug) {
		this.debug= debug;
	}

	/*
	 * Verbose output
	 */
	void setVerbose(boolean verbose) {
		this.verbose= verbose;
	}

	/**
	 * Return a version string w/o the cvs keyword (i.e. $Id$).
	 */
	public static String version() {
		String st= "";
		char[] woId= new char[st.length()];
		int srcBegin= 5;
		int srcEnd= st.length() - 6;
		int srccnt= srcEnd - srcBegin;
		if (srccnt <= 0) {
			return "";
		}
		st.getChars(srcBegin, srcEnd, woId, 0);
		return new String(woId, 0, srccnt);
	}

	/**
	 * Return an iterator for the Lattice.
	 */
	public LatticeIterator latticeIterator() {
		return new LIterator();
	}

	/**
	 * Implementation of iterator support for a lattice.
	 */
	private class LIterator implements LatticeIterator {
		private ListIterator<Element> liter;

		private LIterator() {
			liter = elements.listIterator();
		}

		public boolean hasNext() {
			return liter.hasNext();
		}

		public Element next() {
			return liter.next();
		}

		public void remove() {
			liter.remove();
		}
	}

/**Return the lattice as a DOM document object. */
	public Document getLatticeAsDocument() {
		return new LatticeSynchronizer(this).getDocument();
	}

} ////////////////////// Lattice
